apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1450 @@
+        const CSMWorld::LandHeightsColumn::DataType landRightShapePointer = landTable.data(landTable.getModelIndex(CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY()), landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandHeightsColumn::DataType landDownShapePointer = landTable.data(landTable.getModelIndex(CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY() + 1), landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandNormalsColumn::DataType landNormalsPointer = landTable.data(landTable.getModelIndex(cellId, landnormalsColumn)).value<CSMWorld::LandNormalsColumn::DataType>();
+        CSMWorld::LandNormalsColumn::DataType landNormalsNew(landNormalsPointer);
+
+        // Generate land normals record
+        for(int i = 0; i < ESM::Land::LAND_SIZE; ++i)
+        {
+            for(int j = 0; j < ESM::Land::LAND_SIZE; ++j)
+            {
+                osg::Vec3f v1(128, 0, 0);
+                osg::Vec3f v2(0, 128, 0);
+
+                if (i < ESM::Land::LAND_SIZE - 1) v1.z() = landShapePointer[j * ESM::Land::LAND_SIZE + i + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                else
+                {
+                    std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX() + 1, cellCoordinates.getY());
+                    if (isLandLoaded(cellId))
+                        v1.z() = landRightShapePointer[j * ESM::Land::LAND_SIZE + 1] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                }
+
+                if (j < ESM::Land::LAND_SIZE - 1) v2.z() = landShapePointer[(j + 1) * ESM::Land::LAND_SIZE + i] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                else
+                {
+                    std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY() + 1);
+                    if (isLandLoaded(cellId))
+                        v2.z() = landDownShapePointer[ESM::Land::LAND_SIZE + i] - landShapePointer[j * ESM::Land::LAND_SIZE + i];
+                }
+
+                osg::Vec3f normal = v1 ^ v2;
+                const float hyp = normal.length() / 127.0f;
+
+                normal /= hyp;
+
+                landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 0] = normal.x();
+                landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 1] = normal.y();
+                landNormalsNew[(j * ESM::Land::LAND_SIZE + i) * 3 + 2] = normal.z();
+            }
+        }
+        pushNormalsEditToCommand(landNormalsNew, document, landTable, cellId);
+    }
+    undoStack.endMacro();
+    mAlteredCells.clear();
+}
+
+float CSVRender::TerrainShapeMode::calculateBumpShape(const float& distance, int radius, const float& height)
```

Comment:
Passing float by const reference does not give any benefit. This may even reduce performance because `sizeof(float)` is 32-bit, `sizeof(const float&)` is either 32-bit or 64-bit depending on architecture, so for 64-bit this will take even more memory, but more important reason is that to read this value by reference which is memory address first this address is read from memory and then the value by this address is read from memory, so 2 memory accesses in the worst case. When float is passed by value only a single acceess is required. Also optimization may not even put these values into stack, but into registers directly, so even memory access might not be required. Look at this assembly comparsion: https://godbolt.org/z/X95DE4 (first look at add function assembly then to calculateBumpShape). Probably this place does not affect performance too much but still there is no clear benefit to pass number object by const reference.